### DISCUSSION TITLE
Linking between documents

### DISCUSSION THREAD

**Discussion by jghauser** [papis/papis#530] [Ideas]
Is there a way to link between documents in papis? I've started documenting such links with a `parent_id` key in my `.yaml` files (mostly because I sometimes have chapters that are their own documents for referencing purposes, but have pdfs attached to the main book). In the past, I used to create symlinks for the attachments, but they don't synchronise well. If this doesn't exist, do you think it might make sense to implement? If so do you have any ideas how? I've finally got some more free time on my hand and might take a stab at it :)

**Comment by alexfikl** [comment]
Hm, I don't think there is a way to do that at the moment, since the documents aren't linked in any way. What's the workflow you were going for? The only one I can think of is doing something like
```
papis open <query-for-subchapter>
```
and have it go look for files in the parent if the selected document doesn't have any.

**Comment by jghauser** [comment]
Yes, that's exactly the sort of use case I had in mind. And similarly for the notes. Do you think that would be something useful to implement in papis itself (otherwise, I could do so in my plugin instead)?

Also, can you think of any use cases for linking in the other direction? Should we always have bidrectional links (have a `related-to` key, I think this is what zotero does)? At least for this usecase, bidirectional linking would work just as well. Or should there by `child-of` and `parent-of` keys (or are there relations this would exclude)? Although bidirectional linking would be a bit more tedious to set up, I guess the whole process could be quite easily automated.

**Reply by alexfikl** [reply]
Yeah, it sounds like it could be useful! It should also be fairly noninvasive too: just add a little wrapper function that looks into documents recursively for some keys: `files`, `notes`, `citations` (others?).

`related_to` sounds like a good name, since it lets the user assign meaning to the relationship between the two documents. What would be a nice way to add these fields? Would
```
papis update --set related_to <papis_id> <query>
```
work well enough? 

For future reference, here's what Zotero has on the subject:
https://www.zotero.org/support/related

**Reply by jghauser** [reply]
Yeah, these sounds like the useful keys. And the `papis update` command seems good, too. Great, I'll start looking into how to create this and will get back to you if I have any questions. Thanks for the feedback!

**Reply by jghauser** [reply]
Finally looking at this again and thinking about what to do to get basic functionality to work. `papis update` and `papis add` already allow setting the relevant `related_to` key without any changes. It's not necessarily very ergonomic (because you have to use another command to get the `papis_id` first), but it's good enough for now. We can always find easier ways to do that later.

Seems like we only really need (1) maybe a change to the current open doc/note functions so that they fall back to opening the doc/note specified by `related_to` and (2) document this? Can you think of anything else?

Re (1), should this be behind an option? I would say that it's probably unnecessary because it's only a fallback and will only happen if you have `related_to` explicitly specified. Re (2), where in the docs would be a good place for this?

**Reply by alejandrogallo** [reply]
thinking about this I realised that `papis update` is missing a flag, `--append`, in order to append to the field if there is something else, what do you think?
`papis update --set related_to 6546546 <query> --append`
Also `--listappend` would be a useful addition but only for citations and for tags, and it's questionable if it's worthwhile implementing then.

**Reply by jghauser** [reply]
@alejandrogallo, do you mean `--append` should add a string to an existing string? So, in the present instance, it would be for adding a second `papis_id` so that two documents are mentioned as `related_to`? If so, wouldn't that open the same can of worms as with tags, namely the question of what separator to use etc (although I guess the user could just specify the separator as part of the input string)? In any case, I feel like if we allow multiple docs in `related_to` (which I somehow didn't consider!), it would make more sense to just have that field be a list in the first place. What do you think?

Oh, and if we do lists, can this still be handled by `papis update`?
